### PR TITLE
✨ Clusterctl alpha rollout history

### PR DESCRIPTION
**Pull request by hiromu-a5a**
**What this PR does / why we need it**:

Adds command and client for the clusterctl alpha rollout history.

Although the rollout undo command is already implemented, it's hard to determine a revision number that should be passed to --to-revision option due to the lack of the history command.
The basic behavior of this history command is similar to the kubectl rollout history. A few differences are: i) only the machinedeployment is supported; ii) `clusterctl alpha rollout history --revision=<revision>` **describes** all attributes of a machinset corresponding to that revision **in yaml format**.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Tracking Issue:
https://github.com/kubernetes-sigs/cluster-api/issues/3439

Depends-on:
#7988

### CODE REVIEW

**Comment by hiromu-a5a** [comment]
unhold after #7988 is merged.

File: cmd/clusterctl/client/alpha/rollout_viewer.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		}
+		fmt.Fprint(os.Stdout, string(output))
+		return nil
+	}
+
+	// Print an overview of all revisions
+	// Create a revisionToChangeCause map
+	historyInfo := make(map[int64]string)
+	for _, ms := range msList {
+		v, err := mdutil.Revision(ms)
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/alpha/rollout_viewer.go:81]: I just stumbled into machinedeployment.clusters.x-k8s.io/revision-history and I'm wondering if we should keep this as well into account while building the revision history

Comment (hiromu-a5a) [reply cmd/clusterctl/client/alpha/rollout_viewer.go:81]: Thank you for your comments. 
Maybe, it's a good time to discuss the deprecation of `revision-history`. I personally can't find any use cases for this annotation but we need to confirm that, and I think it should be handled by another issue. 

FYI, I found that Kubernetes also has [this annotation](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/util/deployment_util.go#L50), but it's not used in the actual logic.

**Comment by hiromu-a5a** [comment]
I'll update the note v1.3 to v1.4 later.

File: cmd/clusterctl/client/alpha/rollout_viewer.go (hunk 1 above)
Comment (fabriziopandini) [reply cmd/clusterctl/client/alpha/rollout_viewer.go:81]: In CAPI it is used in [https://github.com/kubernetes-sigs/cluster-api/blob/29618402670e19c77833edbede6d9e[…]7b4d1036a/internal/controllers/machinedeployment/mdutil/util.go](https://github.com/kubernetes-sigs/cluster-api/blob/29618402670e19c77833edbede6d9e87b4d1036a/internal/controllers/machinedeployment/mdutil/util.go#L222-L234)

TL;DR each machineset keeps track of the revision it currently is, and if “reused”, it keeps also track of the revision it was…

File: api/v1beta1/common_types.go [hunk 2]
```diff
@@ -120,6 +120,10 @@ const (
 	// instead of being a source of truth for eventual consistency.
 	// This annotation can be used to inform MachinePool status during in-progress scaling scenarios.
 	ReplicasManagedByAnnotation = "cluster.x-k8s.io/replicas-managed-by"
+
+	// ChangeCauseAnnotation is the annotation set on MachineSets by users to identify the cause of revision changes.
+	// This annotation will be shown when users run `clusterctl alpha rollout history`.
+	ChangeCauseAnnotation = "cluster.x-k8s.io/change-cause"
```
Comment (ykakarap) [inline api/v1beta1/common_types.go:126]: This annotation is never updated, right? So the cause is always "<none>". Is there a follow-up planned to update this annotation? 
If so, FYI, that work could potentially overlap with the in-place propagation work (https://github.com/kubernetes-sigs/cluster-api/issues/7731).

**Review by ykakarap — CHANGES_REQUESTED** [review]

File: cmd/clusterctl/client/alpha/rollout_viewer.go (hunk 1 above)
Comment (ykakarap) [reply cmd/clusterctl/client/alpha/rollout_viewer.go:81]: I would prefer not showing the revision numbers that are listed under revision history annotation, at least as separate entries in the output. 
Since the idea behind the history command is to list the revisions the user can rollback to, it would be best to not show the revision numbers in the revision history annotation as it is not possible to rollback using those revision numbers. 

If we want really want to show those numbers in the output for the sake of having a full history then how about adding a third column called "matching revisions"/"duplicate reivisions", or some better name where we list these revision numbers against target revision numbers?

Thoughts?

Comment (fabriziopandini) [reply cmd/clusterctl/client/alpha/rollout_viewer.go:81]: Why can't we rollback to revision number in revision history?

File: api/v1beta1/common_types.go (hunk 2 above)
Comment (hiromu-a5a) [reply api/v1beta1/common_types.go:126]: I don't have the plan to update this specific annotation directly from the codes. I assume that this annotation is set by users, like [Kubernetes](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/).

Also, I'm okay with propagating this annotation set in a machinedeployment to machinsets within the scope of the label propagation as Kubernetes does a similar behavior (as described below), but I guess it shouldn't be specific for this annotation.

> CHANGE-CAUSE is copied from the Deployment annotation kubernetes.io/change-cause to its revisions upon creation. You can specify theCHANGE-CAUSE message by:

**Review by hiromu-a5a — COMMENTED** [review]
Thank you for the review.
Please kindly review the updates.

**Comment by hiromu-a5a** [comment]
Split commits for ease of review.

File: cmd/clusterctl/client/alpha/rollout_viewer.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/rollout_viewer.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: cmd/clusterctl/client/alpha/rollout_viewer.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+	if revision < 0 {
+		return errors.Errorf("revision number cannot be negative: %v", revision)
+	}
+
+	// Print details of a specific revision
+	if revision > 0 {
+		ms, err := findMachineDeploymentRevision(revision, msList)
+		if err != nil {
+			return errors.Errorf("unable to find the spcified revision")
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/rollout_viewer.go:68]:
```suggestion
			return errors.Errorf("unable to find the specified revision %d for MachineDeployment %s", revision, d.Name)
```

File: cmd/clusterctl/client/alpha/rollout_viewer.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		return nil
+	}
+
+	// Print an overview of all revisions
+	// Create a revisionToChangeCause map
+	historyInfo := make(map[int64]string)
+	for _, ms := range msList {
+		v, err := mdutil.Revision(ms)
+		if err != nil {
+			log.V(7).Error(err, fmt.Sprintf("unable to get revision from machineset %s for machinedeployment %s in namespace %s", ms.Name, d.Name, d.Namespace))
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/rollout_viewer.go:84]: Just out of curiosity, is there a convention for using log level 7 here for non-critical errors?

File: cmd/clusterctl/client/alpha/rollout_viewer_test.go [hunk 6]
```diff
@@ -0,0 +1,338 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/rollout_viewer_test.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: cmd/clusterctl/client/alpha/rollout_viewer_test.go [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+	defer func() {
+		os.Stdout = stdout
+	}()
+	os.Stdout = w
+
+	fnc()
+	_ = w.Close()
+
+	var buf bytes.Buffer
+	_, _ = io.Copy(&buf, r)
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/rollout_viewer_test.go:48]: Do we need to have the `_, _` assignment here?

File: cmd/clusterctl/cmd/rollout/history.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (Jont828) [inline cmd/clusterctl/cmd/rollout/history.go:2]:
```suggestion
Copyright 2023 The Kubernetes Authors.
```

File: go.mod [hunk 9]
```diff
@@ -133,7 +133,7 @@ require (
 	google.golang.org/protobuf v1.28.1 // indirect
 	gopkg.in/inf.v0 v0.9.1 // indirect
 	gopkg.in/ini.v1 v1.67.0 // indirect
-	gopkg.in/yaml.v2 v2.4.0 // indirect
+	gopkg.in/yaml.v2 v2.4.0
```
Comment (Jont828) [inline go.mod:136]: Is there a reason to have yaml.v2 and yaml.v3 both imported?

**Review by Jont828 — COMMENTED** [review]
Thanks for the contribution! This looks pretty good, just a few small comments.

File: cmd/clusterctl/client/alpha/rollout_viewer.go (hunk 1 above)
Comment (sbueringer) [reply cmd/clusterctl/client/alpha/rollout_viewer.go:82]: Following up from the discussion in the office hours.

I think it's very user unfriendly if we do not support revision-history. Today Cluster API only tracks MachineSets that are actually different via separate MachineSet objects. As soon a MachineSet is the same as another one in the history revision-history is used automatically.

If we don't support revision-history from a user point-of-view there will be arbritrary holes in the history (e.g. if 1, 4, 7 are the same the history is: 2, 3, 5, 6, 8). I think this is very hard to understand for users.

I would suggest:
* in the history command: list all revisions including the one from the revision-history annotation
* for the rollback command: make it possible to also rollback to revisions from the revision history (should be as easy as also looking at the revision-history annotation instead of only at the revision annotation to find the MachineSet for a specific revision)

File: go.mod (hunk 9 above)
Comment (hiromu-a5a) [reply go.mod]: Looks like it is already included in the [main branch](https://github.com/kubernetes-sigs/cluster-api/blob/main/go.mod#L136). After I changed `rollout_viewer.go` to use yaml.v3, it is still imported indirectly 🤔 .

File: cmd/clusterctl/client/alpha/rollout_viewer_test.go (hunk 7 above)
Comment (hiromu-a5a) [reply cmd/clusterctl/client/alpha/rollout_viewer_test.go]: Otherwise the following linter error happens, but ignoring err is certainly bad practice... I changed it to return an error. Thank you for pointing it out.

```bash
cmd/clusterctl/client/alpha/rollout_viewer_test.go:48:9: Error return value of `io.Copy` is not checked (errcheck)
        io.Copy(&buf, r)
```

File: cmd/clusterctl/client/alpha/rollout_viewer.go (hunk 5 above)
Comment (hiromu-a5a) [reply cmd/clusterctl/client/alpha/rollout_viewer.go]: I found that Error is always logged regardless of V-level in the first place. I just removed `V(7)`.
As the answer for your question, I think the absence of revision is abnormal in CAPI and thus the clusterctl should always notify users of the error.

File: cmd/clusterctl/client/alpha/rollout_viewer.go (hunk 1 above)
Comment (hiromu-a5a) [reply cmd/clusterctl/client/alpha/rollout_viewer.go:87]: Thank you for following up and My apologies for the delayed reply.

I changed PR following your suggestion. 

I'll deal with your second suggestion in a different PR as it should be an enhancement of `undo`.

> should be as easy as also looking at the revision-history annotation instead of only at the revision annotation to find the MachineSet for a specific revision

This makes sense.

---

BTW, I made output of the history command like this. If I am a user, I want to group revisions associated to the same MachineSet with the same change cause.

```bash
REVISIONS       CHANGE-CAUSE
        2             <none>  
    4,6,8,10      add label
   1,3,5,7,9,11   update to the latest k8s version
```

The above is more useful than this when you are looking for a revision that you want to rollback.

```bash
REVISION        CHANGE-CAUSE
        1             <none>  
        2             <none>  
        3             <none>  
        4             <none>  
        5             <none>  
        6             <none>  
        7             <none>  
        8             <none>  
        9             <none>  
       10            add label  
       11            update to the latest k8s version
```

As an alternative, we can extend the change cause to record its history but I think that is too much.

**Comment by hiromu-a5a** [comment]
I've applied all comments. Please kindly review it again when you have time :)

**Comment by Jont828** [comment]
@hiromu-a5a Looks like you need to rebase. Do you need any help with this PR?

**Comment by hiromu-a5a** [comment]
@Jont828 Sorry for the late reply. I think this PR is now ready for review again. Could you check it?

**Comment by Jont828** [comment]
@hiromu-a5a Are you still working on this? If not we might want to close this PR as it's been without activity for almost 6 months like the other two.

**Comment by sbueringer** [comment]
At this point it's very very likely that we will deprecate and remove the revision management, see https://github.com/kubernetes-sigs/cluster-api/issues/10479

**Comment by hiromu-a5a** [comment]
I am really sorry for the late reply. We are shifting to GitOps and lost motivation to clusterctl.
